### PR TITLE
[mesheryctl] (child) View Performance Profile’s Configuration …

### LINKED ISSUE #2827: [mesheryctl] (child) View Performance Profile’s Configuration
_See [mesheryctl Command Tracker](https://bit.ly/3dqXy1q) for current status of commands._

See epic [perf commands for performance tests](https://github.com/layer5io/meshery/issues/2823) and the [design spec](https://docs.google.com/document/d/1znOqTofsXkcPTF1AXVVSHGEZa5kue6G8vGHM4nsc2tc/edit#) for more details.

#### Desired Behavior
As a `mesheryctl` user, I would like to view the configuration of a performance profile , so that I can decide which profile to use to run the test.

#### Implementation
A new subcommand `mesheryctl perf view <profile name>` which will show the current configuration of the given performance profile.

An optional flag `-o json` will be also available to format the output in json.

#### Acceptance Tests
`mesheryctl perf view <profile name>` should show the configuration of the given profile.

`mesheryctl perf view <profile name> -o json` should output the configuration of a given profile in json format.

#### Mockup

View test configuration of a profile-

```
❯ mesheryctl perf view soak-test

Name- soak-test
Endpoints- https://10.0.0.1/foo
Load generators- fortio
Test run duration- 30s
```

---

#### Contributor Resources

- [mesheryctl Contributing Guide](https://github.com/layer5io/meshery/blob/master/mesheryctl/README.md)
- [Beginner's guide to contributing to Meshery and mesheryctl](https://youtu.be/hh_kFLZx3G4)
- [mesheryctl Command Tracker](https://docs.google.com/spreadsheets/d/1q63sIGAuCnIeDs8PeM-0BAkNj8BBgPUXhLbe1Y-318o/edit#gid=0)
- [Meshery CLI Commands and Documentation](https://docs.google.com/document/d/1xRlFpElRmybJ3WacgPKXgCSiQ2poJl3iCCV1dAalf0k/edit#heading=h.5fucij4hc5wt)

### PR DESCRIPTION
**Pull request by fallmor**
…w/signoff

Signed-off-by: fallmor <fall.mor@outlook.com>

**Description**

This PR fixes #2827

**Notes for Reviewers**

**[Signed commits](https://github.com/layer5io/meshery/blob/master/CONTRIBUTING.md#signing-off-on-commits-developer-certificate-of-origin)**
- [ ] Yes, I signed my commits.
 

``` meshery/mesheryctl on 🌱 prof [!?] on ☁️  (us-east-1) on ☁️ bayemor1603@gmail.com took 3s 
➜ ./mesheryctl perf view mor                
A new release of mesheryctl is available: v0.5.2 → v0.5.13
https://github.com/layer5io/meshery/releases/tag/v0.5.13
Check https://docs.meshery.io/guides/upgrade#upgrading-meshery-cli for instructions on how to update mesheryctl

name: mor1
endpoint: https://git.corp.cloudwatt.com/caascad/monitoring/merge_requests/585
load_generators fortio
Test run duration 30s
#####################
name: mor
endpoint: https://github.com/gophercises/quiz/blob/master/problems.csv
load_generators fortio
Test run duration 30s
#####################
name: mor
endpoint: http://192.168.1.15/productpage
load_generators fortio
Test run duration 30s
#####################
meshery/mesheryctl on 🌱 prof [!?] on ☁️  (us-east-1) on ☁️ bayemor1603@gmail.com 
➜ ./mesheryctl perf view mor -o json        
A new release of mesheryctl is available: v0.5.2 → v0.5.13
https://github.com/layer5io/meshery/releases/tag/v0.5.13
Check https://docs.meshery.io/guides/upgrade#upgrading-meshery-cli for instructions on how to update mesheryctl

{
  "Name": "mor1",
  "Endpoint": "https://git.corp.cloudwatt.com/caascad/monitoring/merge_requests/585",
  "QPS": 300,
  "Duration": "30s",
  "Loadgenerator": "fortio"
}
{
  "Name": "mor",
  "Endpoint": "https://github.com/gophercises/quiz/blob/master/problems.csv",
  "QPS": 300,
  "Duration": "30s",
  "Loadgenerator": "fortio"
}
{
  "Name": "mor",
  "Endpoint": "http://192.168.1.15/productpage",
  "QPS": 300,
  "Duration": "30s",
  "Loadgenerator": "fortio"
}
meshery/mesheryctl on 🌱 prof [!?] on ☁️  (us-east-1) on ☁️ bayemor1603@gmail.com 
➜ ./mesheryctl perf view -h         
See the configuration of your performance profile

Usage:
  mesheryctl perf view [flags]

Examples:
mesheryctl perf view [ performance test profile name ]

Flags:
  -h, --help                   help for view
  -o, --output-format string   (optional) format to display in json

Global Flags:
      --config string   config file (default location is: /Users/morfall/.meshery/config.yaml)
  -v, --verbose         verbose output
meshery/mesheryctl on 🌱 prof [!?] on ☁️  (us-east-1) on ☁️ bayemor
 ```

### CODE REVIEW

**Comment by leecalcote** [comment]
@fallmor fyi, if you list the associated issue in the description the PR after "fixes #", that issue will be automatically closed when the PR is eventually merged - https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword

**Comment by fallmor** [comment]
hello @vijcheru  i think i need some help here with this error. 
`➜ ./mesheryctl perf view mor
A new release of mesheryctl is available: v0.5.2 → v0.5.12
https://github.com/layer5io/meshery/releases/tag/v0.5.12
Check https://docs.meshery.io/guides/upgrade#upgrading-meshery-cli for instructions on how to update mesheryctl

Error: Get "/": stopped after 10 redirects
Usage:
  mesheryctl perf view [flags]

Flags:
  -h, --help   help for view

Global Flags:
      --config string   config file (default location is: /Users/morfall/.meshery/config.yaml)
  -v, --verbose         verbose output
`

i think it's a problem with api

**Comment by leecalcote** [comment]
// @piyushsingariya

File: mesheryctl/internal/cli/root/perf/view.go [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+	Use:   "view",
+	Short: "view perf profile",
+	Long:  `See the configuration of your performance profile`,
+	Args:  cobra.ExactArgs(1),
+	RunE: func(cmd *cobra.Command, args []string) error {
+		mctlCfg, err := config.GetMesheryCtl(viper.GetViper())
+		if err != nil {
+			return errors.Wrap(err, "error processing config")
+		}
+		pattern := args[0]
```
Comment (piyushsingariya) [inline mesheryctl/internal/cli/root/perf/view.go:29]:
```suggestion
		profileID := args[0]
```

File: mesheryctl/internal/cli/root/perf/view.go [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+			return err
+		}
+		client := &http.Client{}
+		resp, err := client.Do(req)
+		if err != nil {
+			return err
+		}
+		if resp.StatusCode != 200 {
+			// failsafe for the case when a valid uuid v4 is not an id of any pattern (bad api call)
+			return errors.Errorf("Response Status Code %d, possible invalid ID", resp.StatusCode)
```
Comment (piyushsingariya) [inline mesheryctl/internal/cli/root/perf/view.go:48]:
```suggestion
			return errors.Errorf("Performance profile `%s` not found. Please verify profile name and try again. Use `mesheryctl perf list` to see a list of performance profiles.", profileID)
```

Comment (fallmor) [reply mesheryctl/internal/cli/root/perf/view.go:48]: yes, you are right. thank you i will fix it

File: mesheryctl/internal/cli/root/perf/view.go [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+	"github.com/spf13/viper"
+)
+
+var outFormatFlag string
+
+var viewCmd = &cobra.Command{
+	Use:     "view",
+	Short:   "view perf profile",
+	Long:    `See the configuration of your performance profile`,
+	Example: "mesheryctl perf view [ performance test profile name ]",
```
Comment (piyushsingariya) [inline mesheryctl/internal/cli/root/perf/view.go:25]:
```suggestion
	Example: "mesheryctl perf view [profile-id]",
```
We're only able to fetch profile based on profile-ids

File: mesheryctl/internal/cli/root/perf/view.go [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+	Args:    cobra.MaximumNArgs(1),
```
Comment (piyushsingariya) [inline mesheryctl/internal/cli/root/perf/view.go:26]:
```suggestion
	Args:    cobra.ExactArgs(1),
```
We always need the performance-id as an argument.

File: mesheryctl/internal/cli/root/perf/view.go [hunk 5]
```diff
@@ -0,0 +1,94 @@
...
+	RunE: func(cmd *cobra.Command, args []string) error {
+		mctlCfg, err := config.GetMesheryCtl(viper.GetViper())
+		if err != nil {
+			return errors.Wrap(err, "error processing config")
+		}
+		proName := args[0]
+		var req *http.Request
+		url := mctlCfg.GetBaseMesheryURL() + "/api/user/performance/profiles"
```
Comment (piyushsingariya) [inline mesheryctl/internal/cli/root/perf/view.go:34]:
```suggestion
		url := mctlCfg.GetBaseMesheryURL() + "/api/user/performance/profiles/" + profileID
```
`view` command is to only fetch one profile's details.

File: mesheryctl/internal/cli/root/perf/view.go [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+		if resp.StatusCode != 200 {
+			// failsafe for the case when a valid uuid v4 is not an id of any pattern (bad api call)
+			return errors.Errorf("Performance profile `%s` not found. Please verify profile name and try again. Use `mesheryctl perf list` to see a list of performance profiles.", proName)
+		}
+		defer resp.Body.Close()
+		data, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return errors.Wrap(err, utils.PerfError("failed to read response body"))
+		}
+		var dat map[string]interface{}
```
Comment (piyushsingariya) [inline mesheryctl/internal/cli/root/perf/view.go:58]: There's a struct in `models` named `PerformanceProfile` use that here.
you can check out the JSON response by sending a GET request at this URL
```
http://localhost:30611/api/user/performance/profiles/:id
```

**Review by piyushsingariya — CHANGES_REQUESTED** [review]

File: mesheryctl/internal/cli/root/perf/view.go (hunk 3 above)
Comment (fallmor) [reply mesheryctl/internal/cli/root/perf/view.go:25]: ok, so the issue was not written correctly ? https://github.com/layer5io/meshery/issues/2827

Comment (piyushsingariya) [reply mesheryctl/internal/cli/root/perf/view.go:25]: Actually, that was discovered after writing that issue. :sweat_smile:

**Comment by leecalcote** [comment]
@fallmor still "WIP" or can we remove this from the PR title and ask @layer5io/mesheryctl-contributors to review?

**Comment by fallmor** [comment]
@leecalcote  yes we can remove the WIP

**Comment by leecalcote** [comment]
@fallmor will you add `-t` as a shorthand flag for `--token`?

@piyushsingariya will you assist in defining a number of unit tests that should be included here? @fallmor getting some unit tests written will save us time as we go to test various scenarios; e.g. user doesn't present a token or there are not perf profiles or perf tests results; or there are 9,000 results, and so on.

If for now those are only tests cases defined in [this test plan](https://docs.google.com/spreadsheets/d/1jDiL2qyT5PrDyMAcoCbbsA8dY0sqfN153J956WJaOYQ/edit#gid=0), that'll work until they are converted into unit tests.

**Comment by piyushsingariya** [comment]
Unit Tests Can be
- No token provided
- No profile name provided
- `perf view -t <path>`
- `perf view --token <path-to-token>`
- Error case if no profile match found for a given profile name
- Format output

@leecalcote the `--token` thing is being taken care of globally in `perf`  #2928 plus `view` command is only covering profiles right now, not results. Should we stretch `view` command to take care of results too?

**Comment by fallmor** [comment]
So if I understand, I need to create a test file to cover all the units test mentioned by @piyushsingariya ? Or I need to just catch the different case on the file wiew.go file ?

**Comment by piyushsingariya** [comment]
@fallmor No, you don't have to create any test files, you just need to make sure that these scenarios are being covered by your code.

Unit Tests are being planned and will be executed after a certain time.

**Comment by navendu-pottekkat** [comment]
@fallmor Actually, you can add a unit test for this command. It is better to handle this side by side.

@piyushsingariya Let us get into this practice of writing tests.

**Comment by fallmor** [comment]
thank you @piyushsingariya  i will submit a  PR shortly

**Comment by fallmor** [comment]
hello @piyushsingariya @navendu-pottekkat  i am not able right to create go test  file. i don't know how to handle the token and the profile name in the go test file

**Comment by leecalcote** [comment]
> hello @piyushsingariya @navendu-pottekkat i am not able right to create go test file. i don't know how to handle the token and the profile name in the go test file

These would need to be supplied as inputs to the tests or to have the tests use the Local Provider (None). The latter seems like the way to go, yeah, @kumarabd?

**Comment by leecalcote** [comment]
> Unit Tests Can be
> 
> * No token provided
> * No profile name provided
> * `perf view profileName -t <path>`
> * `perf view profileName --token <path-to-token>`
> * Error case if no profile match found for a given profile name
> * Format output
> 
> @leecalcote the `--token` thing is being taken care of globally in `perf` #2928 plus `view` command is only covering profiles right now, not results. Should we stretch `view` command to take care of results too?

@fallmor great question. We might need `view` for test results, too. But, as we think on that, let's get this PR merged in the meantime.
